controllers/datadogagent/controller_reconcile_agent.go

```diff
@@ -238,3 +240,79 @@ func (r *Reconciler) generateNodeAffinity(p string, affinity *corev1.Affinity) *
 
 	return affinity
 }
+
+func (r *Reconciler) handleProviders(ctx context.Context) (*map[string]struct{}, error) {
+	if err := r.updateProviderStore(ctx); err != nil {
+		return nil, err
+	}
+
+	if err := r.cleanupUnusedProviders(ctx); err != nil {
+		return nil, err
+	}
+
+	return r.providerStore.GetProviders(), nil
+}
+
+// updateProviderStore recomputes the required providers by making a
+// call to the apiserver for an updated list of nodes
+func (r *Reconciler) updateProviderStore(ctx context.Context) error {
+	nodeList := corev1.NodeList{}
+	err := r.client.List(ctx, &nodeList)
+	if err != nil {
+		return err
+	}
+
+	// recompute providers using updated node list
+	providersList := make(map[string]struct{})
+	for _, node := range nodeList.Items {
+		provider := kubernetes.DetermineProvider(node.Labels)
+		if _, ok := providersList[provider]; !ok {
+			providersList[provider] = struct{}{}
+			r.log.V(1).Info("New provider detected", "provider", provider)
+		}
+	}
+
+	r.providerStore.Reset(providersList)
+
+	return nil
+}
+
+// cleanupUnusedProviders deletes ds/eds from providers that are not
+// present in the provider store. If there are no providers in the
+// provider store, do not delete any ds/eds since that would delete
+// all node agents.
+func (r *Reconciler) cleanupUnusedProviders(ctx context.Context) error {
```

Comment:
nit - maybe `cleanupUnusedDaemonsets` since this is cleaning up DS/EDS resources.